android/src/main/kotlin/com/hypertrack/sdk/flutter/common/Serialization.kt

```diff
@@ -0,0 +1,196 @@
+            }
+        }
+    }
+
+    fun serializeHypertrackError(error: HyperTrackError): Map<String, String> {
+        return mapOf(
+            KEY_TYPE to TYPE_HYPERTRACK_ERROR,
+            KEY_VALUE to error.name
+        )
+    }
+
+    fun serializeDeviceId(deviceId: String): Map<String, Any?> {
+        return mapOf(
+            KEY_TYPE to TYPE_DEVICE_ID,
+            KEY_VALUE to deviceId
+        )
+    }
+
+    fun deserializeDeviceName(name: Map<String, Any?>): Result<String> {
+        return parse(name) {
+            it.assertValue<String>(key = KEY_TYPE, value = TYPE_DEVICE_NAME)
+            it.get<String>(KEY_VALUE).getOrThrow()
+        }
+    }
+
+    fun deserializeAvailability(isAvailable: Map<String, Any?>): Result<Boolean> {
+        return parse(isAvailable) {
+            it.assertValue<String>(key = KEY_TYPE, value = TYPE_AVAILABILITY)
+            it.get<Boolean>(KEY_VALUE).getOrThrow()
+        }
+    }
+
+    fun deserializeGeotagData(map: Map<String, Any?>): Result<Geotag> {
+        return parse(map) {
+            val data = it.get<Map<String, Any?>>(KEY_GEOTAG_DATA).getOrThrow()
+            Geotag(data)
+        }
+    }
+
+    fun <T> parse(
+        source: Map<String, Any?>,
+        parseFunction: (ParsingHandle) -> T
+    ): Result<T> {
+        val handle = ParsingHandle(source);
+        return try {
+            if (handle.exceptions.isEmpty()) {
+                Success(parseFunction.invoke(handle))
+            } else {
+                Failure(ParsingExceptions(source, handle.exceptions))
+            }
+        } catch (e: Exception) {
+            Failure(
+                if (handle.exceptions.isNotEmpty()) {
+                    ParsingExceptions(source, handle.exceptions + e)
+                } else {
+                    e
+                }
+            )
+        }
+    }
+
+    internal class ParsingHandle(
+        private val source: Map<String, Any?>,
+    ) {
+        private val _exceptions = mutableListOf<Exception>()
+        val exceptions: List<Exception> = _exceptions
+
+        inline fun <reified T> get(
+            key: String
+        ): Result<T> {
+            return try {
+                Success(source[key]!! as T)
+            } catch (e: Exception) {
+                Failure(
+                    ParsingException(key, e)
+                        .also {
+                            _exceptions.add(it)
+                        }
+                )
+            }
+        }
+
+        inline fun <reified T> assertValue(
+            key: String,
+            value: Any
+        ) {
+            if (source[key] != value) {
+                _exceptions.add(Exception("Assertion failed: $key != $value"))
+            }
+        }
+    }
+
+    internal data class ParsingExceptions(
+        val source: Any,
+        val exceptions: List<Exception>
+    ) : Throwable(exceptions.joinToString("\n").let {
+        "Invalid input:\n\n${source}\n\n${it}"
+    })
```

Comment:
That's going to be a message propagated outside, so there are too many `\n`s

Reply:
[Serialization.k](https://github.com/hypertrack/sdk-flutter/pull/27/files#diff-7eb8b811d00d555bbe3d3e1a1f4274115bab2c09ba2489eb1a4df556c0f6771c)

We already discussed it :) I somehow convinced you that it is OK :)